src/py21cmmc/emulator.py

```diff
@@ -0,0 +1,196 @@
+        log.info('Find emulator_constants npz file.')
+        if len(os.path.dirname(emu_path)) > 1:
+            all_emulator_numbers = np.load(os.path.dirname(emu_path) + '/emulator_constants.npz')
+        else:
+            all_emulator_numbers = np.load('emulator_constants.npz')
+        log.info('Successfully loaded emulator_constants npz file.')
+        self.zs = all_emulator_numbers['zs']
+        self.limits = all_emulator_numbers['limits']
+        self.zs_cut = self.zs[:60]
+        self.ks_cut = all_emulator_numbers['ks'][1:-3]
+        self.PS_mean = all_emulator_numbers['PS_mean']
+        self.PS_std = all_emulator_numbers['PS_std']
+        self.Tb_mean = all_emulator_numbers['Tb_mean']
+        self.Tb_std = all_emulator_numbers['Tb_std']
+        self.Ts_mean = all_emulator_numbers['Ts_mean']
+        self.Ts_std = all_emulator_numbers['Ts_std']
+        self.uv_lf_zs = np.array([6,7,8,10])
+
+        self.PS_err = all_emulator_numbers["PS_err"]
+        self.Tb_err = all_emulator_numbers["Tb_err"]
+        self.Ts_err = all_emulator_numbers['Ts_err']
+        self.xHI_err = all_emulator_numbers['xHI_err']
+        self.tau_err = all_emulator_numbers['tau_err']
+        log.info('Finished emulator init sucessfully.')
+
+    def predict(self, params):
+        r"""
+        Call the emulator, evaluate it at the given parameters, restore dimensions.
+    
+        Parameters
+        ----------
+        theta : p21.AstroParams or np.ndarray
+        An array with the nine astro_params input all $\in [0,1]$.
+        """
+        log.info('Beginning prediction...')
+        if isinstance(params, p21.AstroParams) or type(params) == dict:
+            log.info('Must convert input params from AstroParams or dict into array.')
+            param_keys = ['F_STAR10','ALPHA_STAR','F_ESC10','ALPHA_ESC','M_TURN', 
+               't_STAR','L_X','NU_X_THRESH','X_RAY_SPEC_INDEX']
+            theta = np.zeros(len(param_keys))
+            for i, key in enumerate(param_keys):
+                if isinstance(params, p21.AstroParams):
+                    theta[i] = params.defining_dict[key]
+                else:
+                    theta[i] = params[key]
+        elif type(params) == np.ndarray:
+            theta = params.copy()
+        else:
+            raise TypeError('theta is in the wrong format. Should be AstroParams object or nine astrophysical parameters in same order as param_keys.')
+        if len(theta.shape) == 1:
+            theta = theta.reshape([1,-1])
+        normed = True
+        # Check that theta is normalized, if not, normalise it.
+        if isinstance(params, p21.AstroParams) or max(theta.ravel()) > 1 or min(theta.ravel()) < 0:
+            log.info('Theta is not normalized.')
+            normed = False # to indicate that input params was not normalised
+            for i in range(theta.shape[-1]):
+                if i == 0 or i == 2 or i == 4 or i == 6:
+                    theta[:,i] = np.log10(theta[:,i])
```

Comment:
double check whether these thetas are already in log space or not

Reply:
Quadruple checked already. The input to AstroParams is in logspace, but not when you read them out of the dict.